### PR TITLE
Integrate Vertex AI Models into GuardRails

### PR DESCRIPTION
**Pull request by aishwaryap**
Update to PR https://github.com/NVIDIA/NeMo-Guardrails/pull/387 (which will be closed)

This PR provides support for calling Vertex AI models in GuardRails. Topical evaluation has been done with `gemini-1.0-pro` and `text-bison` models and self check evaluation for hallucination, fact checking and content moderation has been done with `gemini-1.0-pro`.

Note that https://github.com/GoogleCloudPlatform/generative-ai/issues/344 with Vertex AI models results in calls occasionally erroring out. This PR still contains some very general `try-catch` blocks in hallucination and moderation evaluations to handle this. Errors are reported as a separate category in evaluation.

### CODE REVIEW

File: docs/user_guides/llm/vertexai/config/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+The ABC bot has the following guardrails enabled:
+
+1. Input validation using a [self-check input](../../../docs/user_guides/guardrails-library.md#self-check-input) rail.
+2. Output moderation using a [self-check output](../../../docs/user_guides/guardrails-library.md#self-check-output) rail.
+3. Topical rails, i.e., preventing the bot from talking about unwanted topics, using dialog rails (see [disallow.co](./rails/disallowed.co)).
+
+## Test
+
+To test this configuration, you can use the CLI Chat by running the following command from the `examples/bots/abc` folder:
```
Comment (drazvan) [inline docs/user_guides/llm/vertexai/config/README.md]: This README needs to be updated.

File: examples/configs/llm/vertexai/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# NeMo LLM Service Example
```
Comment (drazvan) [inline examples/configs/llm/vertexai/README.md]: This needs to be updated.

File: docs/user_guides/llm/vertexai/README.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+2. Install the `gcloud` CLI ([guide](https://cloud.google.com/sdk/docs/install)). Note that although 3.8 - 3.12 are listed as supported, [this error](https://stackoverflow.com/questions/77316716/gcloud-modulenotfounderror-no-module-named-imp) occurs on Python 3.12. This guide was tested using Python 3.10.2.
+3. Create a service account following [this guide](https://cloud.google.com/iam/docs/service-accounts-create) and grant it the role of `Vertex AI Service Agent`.
+4. Create and download a service account key for the service account ([guide](https://cloud.google.com/iam/docs/keys-create-delete)).
+5. Enable the Vertex AI API ([guide](https://cloud.google.com/vertex-ai/docs/start/cloud-environment#:~:text=Enable%20Vertex%20AI%20APIs,-In%20the%20Google&text=Click%20Enable%20All%20Recommended%20APIs,the%20APIs%20are%20being%20enabled.))
+6. Install additional python libraries needed to call Vertex AI using `pip install google-cloud-aiplatform>=1.38.0`
```
Comment (drazvan) [inline docs/user_guides/llm/vertexai/README.md]: We should move the steps about enabling and testing(the snippet below with `start_chat()`) VertexAI into a separate guide in `docs/user_guides/advanced`. This guide should mention: "This guide assumes you have configured VertexAI. If not, refer to this guide.". It should only mention installing the dependencies and setting the environment variable.

File: docs/user_guides/llm/vertexai/README.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+The United States has a long and complex history. The first European settlers arrived in the early 17th century. The United States gained independence from Great Britain in 1776. The country has been involved in a number of wars, including the American Civil War, the Spanish-American War, and World War II.
+
+The United States is a diverse country, with people from all over the world. The country has a rich culture, with a wide variety of music, art, and literature. The United States is also a major center of technology and innovation.
+------
+User: Where was its 44th president born?
+Gemini: The 44th President of the United States, Barack Obama, was born in Honolulu, Hawaii on August 4, 1961.
+------
+```
+
+Now we can test Vertex AI models with GuardRails. The `config` directory includes a variation of the ABC Bot but calling the Gemini 1.0 Pro model from Vertex AI. The main difference is in the `config.yml` file. The relevant section is
```
Comment (drazvan) [inline docs/user_guides/llm/vertexai/README.md]: We should not duplicate the config in this folder. The notebook should copy the config when running and update the models part. The config folder itself should also be added to `.gitignore` so it doesn't get committed.

**Review by drazvan — CHANGES_REQUESTED** [review]

**Comment by drazvan** [comment]
@trebedea : can you please review the changes to the evaluation part? Thanks!

File: docs/user_guides/llm/vertexai/config/README.md (hunk 1 above)
Comment (aishwaryap) [reply docs/user_guides/llm/vertexai/config/README.md]: Updated in commit [5319dbc5b89201e1e4d7046386183dbd7c3b4a12](https://github.com/NVIDIA/NeMo-Guardrails/pull/402/commits/5319dbc5b89201e1e4d7046386183dbd7c3b4a12).

File: examples/configs/llm/vertexai/README.md (hunk 2 above)
Comment (aishwaryap) [reply examples/configs/llm/vertexai/README.md]: Updated in commit [5319dbc5b89201e1e4d7046386183dbd7c3b4a12](https://github.com/NVIDIA/NeMo-Guardrails/pull/402/commits/5319dbc5b89201e1e4d7046386183dbd7c3b4a12).

File: docs/user_guides/llm/vertexai/README.md (hunk 3 above)
Comment (aishwaryap) [reply docs/user_guides/llm/vertexai/README.md]: Updated in commit [5319dbc5b89201e1e4d7046386183dbd7c3b4a12](https://github.com/NVIDIA/NeMo-Guardrails/pull/402/commits/5319dbc5b89201e1e4d7046386183dbd7c3b4a12).

File: docs/user_guides/llm/vertexai/README.md (hunk 4 above)
Comment (aishwaryap) [reply docs/user_guides/llm/vertexai/README.md]: Updated in commit [5319dbc5b89201e1e4d7046386183dbd7c3b4a12](https://github.com/NVIDIA/NeMo-Guardrails/pull/402/commits/5319dbc5b89201e1e4d7046386183dbd7c3b4a12).

File: nemoguardrails/eval/evaluate_moderation.py [hunk 5]
```diff
@@ -224,6 +242,10 @@ def run(self):
             print(
                 f"% of samples correctly flagged by output moderation rail: {check_output_correct/len(self.dataset) * 100}"
             )
+            if check_output_error > 0:
+                print(
+                    f"% of samples correctly flagged by output moderation rail: {check_output_error/len(self.dataset) * 100}"
```
Comment (trebedea) [inline nemoguardrails/eval/evaluate_moderation.py]:
```suggestion
                    f"% of samples where output moderation model or rail errored out: {check_output_error/len(self.dataset) * 100}"
```

File: nemoguardrails/eval/evaluate_moderation.py [hunk 6]
```diff
@@ -213,6 +227,10 @@ def run(self):
             print(
                 f"% of samples correctly flagged by jailbreak rail: {jailbreak_correct/len(self.dataset) * 100}"
             )
+            if jailbreak_error > 0:
+                print(
+                    f"% of samples where model or rail errored out: {jailbreak_error/len(self.dataset) * 100}"
```
Comment (trebedea) [inline nemoguardrails/eval/evaluate_moderation.py]:
```suggestion
                    f"% of samples where jailbreak model or rail errored out: {jailbreak_error/len(self.dataset) * 100}"
```

File: nemoguardrails/llm/providers/providers.py [hunk 7]
```diff
@@ -228,6 +228,20 @@ def get_llm_provider(model_config: Model) -> Type[BaseLanguageModel]:
                 "Could not import langchain_openai, please install it with "
                 "`pip install langchain-openai`."
             )
+
+    elif model_config.engine == "vertexai":
```
Comment (trebedea) [inline nemoguardrails/llm/providers/providers.py:232]: This is tricky, it solves the warning for langchain <0.2.0, but if in any future langchain version they decide the provider for VertexAI will be a different class we will still use the one for 0.2.0. 

On my side, it's better not to manually set providers as langchain does that automatically for their library and they maintain this better than us.

**Review by trebedea — APPROVED** [review]
Looks good, just some minor changes and a comment on Langchain usage.

File: nemoguardrails/llm/providers/providers.py (hunk 7 above)
Comment (drazvan) [reply nemoguardrails/llm/providers/providers.py:232]: I think this solution is ok. This is mainly due to this intermediate phase where LangChain is migrating from providing the LLM providers from the main library vs. individual packages. The former is being deprecated. In the future, the import from the main library will not be available anyway.

**Review by drazvan — APPROVED** [review]

**Comment by drazvan** [comment]
Thanks @aishwaryap ! I've made a few additional changes. Looks good 👍 !
I'll merge this as is right now, for future contributions make sure the commits are signed: https://github.com/NVIDIA/NeMo-Guardrails/blob/main/CONTRIBUTING.md#submitting-your-work.

**Pull request merged**